linalg.norm

gboduljak opened the pull request:
## Proposed changes

Implemented a method for computing matrix/vector norms. Analogous to `numpy.linalg.norm`.

## Checklist

Put an `x` in the boxes that apply.

- [x] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx/blob/main/CONTRIBUTING.md) document
- [x] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the necessary documentation (if needed)

gboduljak commented:
I figured this out.

awni commented:
@gboduljak this looks really nice!  Regarding API we are trying to stick to NumPy unless we have a really good reason to diverge. So could you please rename `vector_norm` to `norm`?

awni commented on mlx/linalg.cpp:
```diff
@@ -0,0 +1,64 @@
+          [&](double p) {
+            if (p >= 1)
+              return power(
+                  sum(power(abs(a, s), array(p), s), axes, keepdims, s),
```
I would special case `p==2` here because you can do `sqrt(sum(square(a, s)))` it will be faster and it's by far the most common case.

awni replied:
Same comment for `p==1` special case `sum(abs(...`

gboduljak replied:
I agree. Thanks for pointing this out. I will implement as you suggested.

awni commented on mlx/linalg.cpp:
```diff
@@ -0,0 +1,64 @@
+              return sum(
+                  where(a != 0, array(1), array(0), s), axes, keepdims, s);
+            else
+              throw std::invalid_argument(
+                  "[core.linalg.norm] p norm is defined only for p >= 1.");
```
Why not support `0 < p < 1`?

gboduljak replied:
https://statisticaloddsandends.wordpress.com/2020/05/27/lp-norm-is-not-a-norm-when-p-1/

awni replied:
Yes it's technically not a norm, nevertheless it is supported by numpy and commonly used for sparsity. Since it's a trivial modification, I'd be in favor of adding it.

gboduljak replied:
I will support 0 < p < 1. :)

awni commented on mlx/linalg.h:
```diff
@@ -0,0 +1,45 @@
+    Otherwise, the norm is computed over axes and the other dimensions are
+treated as batch dimensions.
+*/
+array vector_norm(
+    const array& a,
+    const std::variant<double, std::string>& ord = 2.0,
+    const std::vector<int>& axes = {},
+    bool keepdims = false,
+    StreamOrDevice s = {});
+array vector_norm(
+    const array& a,
+    const std::variant<double, std::string>& ord = 2.0,
+    bool keepdims = false,
+    StreamOrDevice s = {});
```
I would prefer you do overloads for `const string& ord` and `double ord` rather than use a variant. There is no reason to do the dispatch at runtime when using the C++ API  so we should just have the overloads. I also think the variant makes the code a little cumbersome / harder to maintain.

Note you will have to update the python binding to do the dispatch instead.

gboduljak replied:
I will switch to overloads instead of the variant.

awni reviewed (changes requested):
This is a really nice addition! I would love to add it, just requested a few changes. Let me know when you are ready to re-review.

gboduljak commented:
> @gboduljak this looks really nice! Regarding API we are trying to stick to NumPy unless we have a really good reason to diverge. So could you please rename `vector_norm` to `norm`?

Yes, I can rename it to `norm`.  For now, I kept it as `vector_norm` because it supports only vector normalization. 
My concern is that for now, we cannot maintain full compatibility with `np.norm`. This is because we are missing nuclear norm and other norms that depend on the spectrum. Do we want to merge after renaming to `norm` even if we are not fully compatible with `np.norm`? 

I do not think it is a good idea to implement spectrum-based norms in this PR because someone is already working on SVD. We can reuse SVD implementation for these norms. We should, however, leave other matrix norms as future work. I also have to implement Frobenious norm. 

Thanks for the feedback :)

awni commented:
@gboduljak no worries the best way to deal with that is to throw for the cases we don't yet support. That's totally reasonable and it's totally reasonable to build out the API in steps as long as we give a helpful error message for not yet implemented cases.

awni commented:
@gboduljak what's the status of this PR? Would love to get `mlx.linalg.norm` in, let me know if you are still planning to work on it.

gboduljak commented:
> @gboduljak what's the status of this PR? Would love to get `mlx.linalg.norm` in, let me know if you are still planning to work on it.

I am working on this right now. I am confident this will be complete tomorrow.

gboduljak commented:
@awni I made some progress on this PR. I implemented missing cases in C++, wrote extensive tests in C++, and updated Python bindings to match the requested changes.  However, we are missing extensive tests for Python bindings, docs, and examples. Please, take a look at the C++ logic, tests, and bindings. I will add Python tests, docs, and examples later today.

gboduljak commented:
@awni I am currently struggling to determine the best possible solution for Python bindings. The problem is that there are at most `3 * 3 * 2 = 18` possible combinations of norm parameters. Assuming we want to avoid runtime dispatch in bindings, we need to account for cases when the user deliberately sets `ord=None` or `axis=None` in the function call. If we want to be fully compatible with `numpy`, we need to support this. However, I am not sure how to support this without C++ runtime dispatch in bindings. In addition, given the complexity of `norm`, we should avoid dispatch in C++, but perhaps we can use runtime dispatch in Python bindings. What do you think?

nullhook commented:
waiting for this to be merged as i have an incoming PR that also introduces linalg namespace for QR factorization

nullhook commented on tests/linalg_tests.cpp:
```diff
@@ -0,0 +1,306 @@
+using namespace mlx::core;
+using namespace mlx::core::linalg;
+
+TEST_CASE("[mlx.core.linalg.norm] no ord") {
```
this file eventually is going to be long when adding other linalg tests. i have an upcoming PR that will potentially add various cases to test QR factorization.

should lingalg test files be prefixed along with it's operation name? for ex: `linalg_norm_tests.cpp`
`linalg_qr_tests.cpp`
`linalg_svd_tests.cpp`

gboduljak replied:
I agree. I think we should separate the tests when necessary. The similar holds for `test_nn` for example.

awni commented:
> In addition, given the complexity of norm, we should avoid dispatch in C++, but perhaps we can use runtime dispatch in Python bindings

@gboduljak it's impossible to avoid runtime dispatch via python so don't worry about that! You can just have cases to choose the relevant function within the binding. That's totally reasonable. The other alternative is overloads in the bindings (which we don't usually do), but the dispatch in that case is still done at run-time but just managed by some pybind11 machinery.

awni commented:
PS this is looking quite impressive! I will try and get it reviewed today!

awni commented on mlx/utils.cpp:
```diff
@@ -279,4 +286,10 @@ std::ostream& operator<<(std::ostream& os, const std::vector<size_t>& v) {
   return os;
 }
 
+std::vector<int> get_shape_reducing_over_all_axes(int ndim) {
+  std::vector<int> shape(ndim);
+  std::iota(shape.begin(), shape.end(), 0);
+  return shape;
+}
```
This seems unused?

gboduljak replied:
Indeed, I replaced it with ``get_reduce_axes``. I should have removed this :(

awni commented on mlx/linalg.cpp:
```diff
@@ -0,0 +1,157 @@
+    return max(abs(a, s), axis, keepdims, s);
+  else if (ord == "-inf")
+    return min(abs(a, s), axis, keepdims, s);
+  std::stringstream error_stream;
+  error_stream << "Invalid ord value " << ord;
```
Use `std::ostringstream` here for consistency with the rest core. (Same below)

awni commented on mlx/linalg.cpp:
```diff
@@ -0,0 +1,157 @@
+// Copyright © 2023 Apple Inc.
+
+#include <iostream>
```
Remove unused include

gboduljak replied:
Good point. I used this for debugging and forgot to remove.

awni commented on mlx/linalg.cpp:
```diff
@@ -0,0 +1,157 @@
+#include "mlx/array.h"
+#include "mlx/linalg.h"
+#include "mlx/ops.h"
+#include "utils.h"
```
Consistency here please `mlx/utils.h`. (I know we are not consistent in general, but let's standardize to the full path for core includes from now on.

gboduljak replied:
Of course. We should replace with 'mlx/utils.h'

awni commented on python/src/overloaded.h:
```diff
@@ -0,0 +1,8 @@
+// Copyright © 2023 Apple Inc.
+
+template <typename... Ts>
+struct overloaded : Ts... {
+  using Ts::operator()...;
+};
+template <class... Ts>
+overloaded(Ts...) -> overloaded<Ts...>;
```
Does this need to be in a separate header? Can it be in linalg.cpp directly?

gboduljak replied:
Maybe we want to reuse this for 'pattern matching'. This is why I went for `overloaded.h`. If you think it not worth having it as a separate header, we can move to `linalg.cpp`.

awni commented on python/src/linalg.cpp:
```diff
@@ -0,0 +1,218 @@
+        return std::visit(
+            overloaded{
+                [&](const double p) {
+                  if (std::isinf((float)p) || std::isinf(p)) {
```
Is the cast to float here with the first `isinf` necessary?

gboduljak replied:
I am not sure. Ideally, we should handle infinity for every 'float' data type, regardless of the precision. I think that Python bindings map infinity to float32 infinity.

awni commented on python/src/linalg.cpp:
```diff
@@ -0,0 +1,218 @@
+        return std::visit(
+            overloaded{
+                [&](const double p) {
+                  if (std::isinf((float)p) || std::isinf(p)) {
+                    if (p > 0) {
+                      return norm(
+                          a,
+                          "inf",
+                          std::holds_alternative<std::monostate>(axis)
+                              ? std::vector<int>()
+                              : get_reduce_axes(axis, a.ndim()),
+                          keepdims,
+                          stream);
+                    }
+                    return norm(
+                        a,
+                        "-inf",
+                        get_reduce_axes(axis, a.ndim()),
+                        keepdims,
+                        stream);
+                  }
```
Also this case handling should be in the C++ `norm`. As in it should support `std::numeric_limits<double>::infinity` (and negative inf). Then you don't need to do these cases for python and we get C++ support.

gboduljak replied:
Ideally, we can remove this case handling and use std::numeric_limits<double>::infinity in C++ code. However, what happens with std::numeric_limits<float>::infinity or  std::numeric_limits<bfloat16>::infinity? I think we should handle or 'infinities' in the same way. This is why I went for the string approach. If you think we should handle only std::numeric_limits<double>::infinity, we can move this to C++ and remove from bindings.

awni commented on python/src/linalg.cpp:
```diff
@@ -0,0 +1,218 @@
+    For values of ``ord < 1``, the result is, strictly speaking, not a
+    mathematical 'norm', but it may still be useful for various numerical
+    purposes.
+
```
This docstring should be in the same style as our other docstrings. See the docstrings in src/ops.cpp for example

awni reviewed (changes requested):
This is looking really nice, I left a few comments to start. I am happy to push some changes myself to keep style stuff consistent rather than badger you with a bunch of nits. Let me know if that's ok!

gboduljak commented on mlx/linalg.h:
```diff
@@ -0,0 +1,31 @@
+
+#pragma once
+
+#include <variant>
```
Good point.

gboduljak commented:
> This is looking really nice, I left a few comments to start. I am happy to push some changes myself to keep style stuff consistent rather than badger you with a bunch of nits. Let me know if that's ok!

I will not be able to address some of your requested changes until Tuesday. However, it is completely fine for me that you take care of such stuff. If you need this before, please implement what is necessary and merge :)

gboduljak commented on mlx/linalg.h:
```diff
@@ -0,0 +1,31 @@
+#include "device.h"
+#include "ops.h"
+#include "stream.h"
+#include "string.h"
```
I had some debugging utils. I should have removed this but I did not. Sorry for that.

gboduljak commented:
@awni I addressed the requested changes for unused imports. However, I am not sure how should we handle infinities for different floating point types. I will be able to finish this on Tuesday. 
I noticed that I left quite a few unused imports in this PR. How can I detect this better in the future? Is there any IDE-plugin or code editor plugin you use for this?

awni commented:
Thanks for making the changes. Will review shortly!

> How can I detect this better in the future? Is there any IDE-plugin or code editor plugin you use for this?

It may exist, but I don't know of it for C++ (would love to know about it if you find something!). One thing you can do is just check your PR after you make it for these sorts of things.

nullhook commented:
use clangd vscode extension, it will notify you about unused includes and have symbol info. you will need a compile_commands.json file in your project's root dir, which is generated by clang.

before running cmake set this:
`export CMAKE_EXPORT_COMPILE_COMMANDS=1`

<img width="279" alt="image" src="https://github.com/ml-explore/mlx/assets/8665427/db1e7d8e-0ed4-4ec8-be28-133bd0de4946">

awni commented:
@gboduljak I pushed an update, make sure you pull before changing anything!

gboduljak commented:
@awni What else we need to do to merge this PR?

gboduljak commented:
> @awni What else we need to do to merge this PR?

I will work on this today and tomorrow.

awni commented:
@gboduljak I think it's mostly good to go? I want to do one more quick pass to make sure we didn't miss anything but o/w I am ready to merge.

gboduljak commented on mlx/linalg.cpp:
```diff
@@ -78,13 +87,13 @@ inline array matrix_norm(
     StreamOrDevice s) {
   if (ord == "f" || ord == "fro") {
-    return sqrt(sum(abs(a, s) * abs(a, s), axis, keepdims, s));
+    return sqrt(sum(square(a, s), axis, keepdims, s), s);
```
If we want to support complex numbers, we might want to replace `square(a,s)` with `a times conjugate(a)`.

awni replied:
It's a good point, I didn't think of that. We should probably modify to branch on complex types and do the abs * abs thing for now if complex (we don't have conjugate yet).

gboduljak replied:
I implemented branching on complex types.

gboduljak commented:
> > How can I detect this better in the future? Is there any IDE-plugin or code editor plugin you use for this?
> 
> use clangd vscode extension, it will notify you about unused includes and have symbol info. you will need a compile_commands.json file in your project's root dir, which is generated by clang.
> 
> before running cmake set this: `export CMAKE_EXPORT_COMPILE_COMMANDS=1`
> 
> <img alt="image" width="279" src="https://private-user-images.githubusercontent.com/8665427/292902444-db1e7d8e-0ed4-4ec8-be28-133bd0de4946.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDM2MzY1NjAsIm5iZiI6MTcwMzYzNjI2MCwicGF0aCI6Ii84NjY1NDI3LzI5MjkwMjQ0NC1kYjFlN2Q4ZS0wZWQ0LTRlYzgtYmUyOC0xMzNiZDBkZTQ5NDYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQUlXTkpZQVg0Q1NWRUg1M0ElMkYyMDIzMTIyNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyMzEyMjdUMDAxNzQwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9Zjg2ZjI3MTM3MzhmMDVjNzM1OTk1ODdmZTEwNDJlYTQ0MjY2YTEwNzBjODNmMmRkOWI2YzczZGVjNjI5NjM2ZCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.Gkhcylou6JgaGoCnwT-63AKOm7wXa_Cl1UMB-m2Ej6w">

Thank you for this suggestion. I will use it in future.

gboduljak commented:
> @gboduljak I think it's mostly good to go? I want to do one more quick pass to make sure we didn't miss anything but o/w I am ready to merge.

@awni I think we are good to go. The only concern I had was about frobenius norm for complex-valued matrices. I addressed this with the latest commit.

awni reviewed (approved):
🚀  Thanks for putting this one together @gboduljak! Great started on a mx.linalg package!
